File: bindings/zig/ove/src/fs.zig
// Copyright (C) 2026 Kamil Lulko <kamil.lulko@gmail.com>
//
// SPDX-License-Identifier: GPL-3.0-or-later
//
// This file is part of oveRTOS.

//! Virtual filesystem — `File` with open/read/write/seek/close and `Dir` for
//! directory enumeration, plus `mount`/`unmount`/`unlink`/`rename` helpers.
//!
//! Wraps `ove/fs.h`. Backed by `littlefs` on FreeRTOS, Zephyr's `fs_*` API on
//! Zephyr, the host POSIX VFS on POSIX, and NuttX's VFS on NuttX. Available
//! when `CONFIG_OVE_FS` is enabled.

const std = @import("std");
const c = @import("c.zig").raw;
const err = @import("error.zig");
const Error = err.Error;

/// Open flag: open for reading.
pub const O_READ = c.OVE_FS_O_READ;
/// Open flag: open for writing.
pub const O_WRITE = c.OVE_FS_O_WRITE;
/// Open flag: create the file if it does not exist.
pub const O_CREATE = c.OVE_FS_O_CREATE;
/// Open flag: append writes to the end of the file.
pub const O_APPEND = c.OVE_FS_O_APPEND;

/// Seek origin: seek relative to the start of the file.
pub const SEEK_SET = c.OVE_FS_SEEK_SET;
/// Seek origin: seek relative to the current file position.
pub const SEEK_CUR = c.OVE_FS_SEEK_CUR;
/// Seek origin: seek relative to the end of the file.
pub const SEEK_END = c.OVE_FS_SEEK_END;

const DIRENT_NAME_LEN = @typeInfo(@TypeOf(@as(c.struct_ove_dirent, undefined).name)).array.len;

/// A single directory entry returned by `Dir.readEntry()`.
pub const Dirent = struct {
    /// Null-terminated entry name (file or subdirectory name, not full path).
    name: [DIRENT_NAME_LEN]u8,
    /// `true` if this entry is a directory, `false` if it is a file.
    is_dir: bool,
    /// File size in bytes (0 for directories).
    size: usize,
};

/// File handle for reading, writing, and seeking within a file.
pub const File = struct {
    handle: c.ove_file_t,

    /// Open a file at `path` with the given `flags` (combination of `O_*` constants).
    ///
    /// Returns a `File` handle on success, or `Error` if the file cannot be opened.
    pub fn open(path: [*:0]const u8, flags: c_int) Error!File {
        var h: c.ove_file_t = null;
        try err.fromCode(c.ove_fs_open(&h, path, flags));
        return .{ .handle = h };
    }

    /// Close the file and release associated resources.
    ///
    /// Sets `handle` to null. Safe to call on an already-closed file.
    pub fn close(self: *File) void {
        if (self.handle != null) {
            _ = c.ove_fs_close(self.handle);
            self.handle = null;
        }
    }

    /// Read up to `buf.len` bytes from the current file position into `buf`.
    ///
    /// Returns the number of bytes actually read. A return value of 0 indicates
    /// end-of-file. Returns `Error` on I/O failure.
    pub fn read(self: File, buf: []u8) Error!usize {
        var bytes_read: usize = 0;
        try err.fromCode(c.ove_fs_read(self.handle, buf.ptr, buf.len, &bytes_read));
        return bytes_read;
    }

    /// Write `buf` to the file at the current position.
    ///
    /// Returns the number of bytes actually written. Returns `Error` on I/O failure
    /// or if the filesystem is out of space.
    pub fn write(self: File, buf: []const u8) Error!usize {
        var bytes_written: usize = 0;
        try err.fromCode(c.ove_fs_write(self.handle, buf.ptr, buf.len, &bytes_written));
        return bytes_written;
    }

    /// Return the total size of this file in bytes.
    ///
    /// Returns `Error` on I/O failure.
    pub fn size(self: File) Error!usize {
        var out: usize = 0;
        try err.fromCode(c.ove_fs_size(self.handle, &out));
        return out;
    }

    /// Move the file position to `offset` bytes relative to `whence`.
    ///
    /// `whence` is one of `SEEK_SET`, `SEEK_CUR`, or `SEEK_END`.
    /// Returns `Error` if the seek position is invalid or the operation fails.
    pub fn seek(self: File, offset: i64, whence: c_int) Error!void {
        try err.fromCode(c.ove_fs_seek(self.handle, @intCast(offset), whence));
    }

    /// Return the current file position as a byte offset from the start.
    pub fn tell(self: File) i64 {
        return c.ove_fs_tell(self.handle);
    }

    // ----- std.io.GenericReader / GenericWriter integration -----

    const FileReader = std.io.GenericReader(File, Error, struct {
        fn read(self: File, buf: []u8) Error!usize {
            return File.read(self, buf);
        }
    }.read);

    const FileWriter = std.io.GenericWriter(File, Error, struct {
        fn write(self: File, buf: []const u8) Error!usize {
            return File.write(self, buf);
        }
    }.write);

    /// `std.io.GenericReader` view of this file.  `Reader.Error =
    /// ove.Error` since the substrate FS layer can fail at runtime
    /// (I/O errors, EOF signalled via `Ok(0)`).
    pub fn reader(self: File) FileReader {
        return .{ .context = self };
    }

    /// `std.io.GenericWriter` view of this file.
    pub fn writer(self: File) FileWriter {
        return .{ .context = self };
    }
};

/// Directory handle for iterating directory entries.
pub const Dir = struct {
    handle: c.ove_dir_t,

    /// Open the directory at `path`.
    ///
    /// Returns a `Dir` handle on success, or `Error` if the path does not exist
    /// or is not a directory.
    pub fn open(path: [*:0]const u8) Error!Dir {
        var h: c.ove_dir_t = null;
        try err.fromCode(c.ove_fs_opendir(&h, path));
        return .{ .handle = h };
    }

    /// Close the directory and release associated resources.
    ///
    /// Sets `handle` to null. Safe to call on an already-closed directory.
    pub fn close(self: *Dir) void {
        if (self.handle != null) {
            _ = c.ove_fs_closedir(self.handle);
            self.handle = null;
        }
    }

    /// Read the next directory entry.
    ///
    /// Returns `null` when all entries have been enumerated (end of directory).
    /// Returns a `Dirent` describing the next entry, or `Error` on I/O failure.
    pub fn readEntry(self: Dir) Error!?Dirent {
        var raw_entry: c.struct_ove_dirent = undefined;
        try err.fromCode(c.ove_fs_readdir(self.handle, &raw_entry));
        // rc == 0 with empty name means end of directory
        if (raw_entry.name[0] == 0) return null;
        var entry: Dirent = .{
            .name = undefined,
            .is_dir = raw_entry.is_dir != 0,
            .size = raw_entry.size,
        };
        @memcpy(&entry.name, &raw_entry.name);
        return entry;
    }
};

/// Mount a filesystem at `mount_point`, backed by the block device at `dev_path`.
///
/// Returns `Error` if the mount fails (e.g. device not found, wrong filesystem type).
pub fn mount(dev_path: [*:0]const u8, mount_point: [*:0]const u8) Error!void {
    try err.fromCode(c.ove_fs_mount(dev_path, mount_point));
}

/// Unmount the filesystem at `mount_point`.
///
/// Flushes pending writes and releases the mount. The mount point becomes
/// inaccessible until `mount()` is called again.
pub fn unmount(mount_point: [*:0]const u8) void {
    c.ove_fs_unmount(mount_point);
}

/// Delete the file at `path`.
///
/// Returns `Error` if the file does not exist or cannot be deleted.
pub fn unlink(path: [*:0]const u8) Error!void {
    try err.fromCode(c.ove_fs_unlink(path));
}

/// Rename (or move) the file at `old_path` to `new_path`.
///
/// Returns `Error` if the source does not exist or the rename fails.
pub fn rename(old_path: [*:0]const u8, new_path: [*:0]const u8) Error!void {
    try err.fromCode(c.ove_fs_rename(old_path, new_path));
}

File: bindings/rust/ove/src/fs.rs
// Copyright (C) 2026 Kamil Lulko <kamil.lulko@gmail.com>
//
// SPDX-License-Identifier: GPL-3.0-or-later
//
// This file is part of oveRTOS.

//! Filesystem abstraction for oveRTOS.
//!
//! Provides RAII [`File`] and [`Dir`] handles for reading, writing, and
//! directory traversal. All paths must be null-terminated byte slices
//! (e.g. `b"/data/config.bin\0"`).

use crate::bindings;
use crate::error::{Error, Result};

// SAFETY (module-wide contract for the `unsafe { bindings::ove_*(...) }` FFI
// calls below): any handle passed to the C API is non-null and refers to a
// live RTOS object — wrapper constructors establish validity via
// `Error::from_code`, and `Drop` (or an explicit `deinit`) is the only place
// a handle is released. Pointer and slice arguments reference caller-owned
// memory valid for the duration of the call; the C side copies whatever it
// retains and does not alias them past return (verified against the
// signatures in `include/ove/*.h`). Blocks that deviate — `transmute`, raw
// pointer casts from user data, slice reconstruction via `from_raw_parts`,
// or storing a callback across the FFI boundary — carry their own
// `// SAFETY:` comment.

/// Open flag: open file for reading.
pub const O_READ: i32 = bindings::OVE_FS_O_READ as i32;
/// Open flag: open file for writing.
pub const O_WRITE: i32 = bindings::OVE_FS_O_WRITE as i32;
/// Open flag: create the file if it does not exist.
pub const O_CREATE: i32 = bindings::OVE_FS_O_CREATE as i32;
/// Open flag: all writes append to the end of the file.
pub const O_APPEND: i32 = bindings::OVE_FS_O_APPEND as i32;

/// Mount a filesystem.
///
/// Pass `None` for `dev` or `mnt` to use platform defaults.
pub fn mount(dev: Option<&[u8]>, mnt: Option<&[u8]>) -> Result<()> {
    let dev_ptr = dev.map_or(core::ptr::null(), |d| d.as_ptr() as *const _);
    let mnt_ptr = mnt.map_or(core::ptr::null(), |m| m.as_ptr() as *const _);
    let rc = unsafe { bindings::ove_fs_mount(dev_ptr, mnt_ptr) };
    Error::from_code(rc)
}

/// RAII file handle.
pub struct File {
    handle: bindings::ove_file_t,
}

impl File {
    /// Open a file. `path` must be `\0`-terminated.
    pub fn open(path: &[u8], flags: i32) -> Result<Self> {
        let mut handle: bindings::ove_file_t = core::ptr::null_mut();
        let rc = unsafe { bindings::ove_fs_open(&mut handle, path.as_ptr() as *const _, flags) };
        Error::from_code(rc)?;
        Ok(Self { handle })
    }

    /// Read from the file into `buf`. Returns the number of bytes read.
    pub fn read(&self, buf: &mut [u8]) -> Result<usize> {
        let mut bytes_read: usize = 0;
        let rc = unsafe {
            bindings::ove_fs_read(
                self.handle,
                buf.as_mut_ptr() as *mut _,
                buf.len(),
                &mut bytes_read,
            )
        };
        Error::from_code(rc)?;
        Ok(bytes_read)
    }

    /// Write `buf` to the file. Returns the number of bytes written.
    pub fn write(&self, buf: &[u8]) -> Result<usize> {
        let mut bytes_written: usize = 0;
        let rc = unsafe {
            bindings::ove_fs_write(
                self.handle,
                buf.as_ptr() as *const _,
                buf.len(),
                &mut bytes_written,
            )
        };
        Error::from_code(rc)?;
        Ok(bytes_written)
    }
}

impl Drop for File {
    fn drop(&mut self) {
        unsafe { bindings::ove_fs_close(self.handle) };
    }
}

// SAFETY: File wraps an opaque RTOS handle. Access is single-threaded by
// application design (only one thread owns the File at a time).
unsafe impl Send for File {}

/// A directory entry returned by `Dir::read_entry`.
pub struct DirEntry {
    inner: bindings::ove_dirent,
}

impl DirEntry {
    /// The entry name as a byte slice (without trailing `\0`).
    pub fn name(&self) -> &[u8] {
        // SAFETY: `self.inner.name` is an inline fixed-size array owned by this
        // `DirEntry`; the slice borrows it for `&self`'s lifetime.
        let name_bytes = unsafe {
            core::slice::from_raw_parts(
                self.inner.name.as_ptr() as *const u8,
                self.inner.name.len(),
            )
        };
        cstr_to_slice(name_bytes)
    }

    /// The file size in bytes.
    pub fn size(&self) -> u32 {
        self.inner.size
    }

    /// Whether this entry is a directory.
    pub fn is_dir(&self) -> bool {
        self.inner.is_dir != 0
    }
}

/// RAII directory handle.
pub struct Dir {
    handle: bindings::ove_dir_t,
}

impl Dir {
    /// Open a directory. `path` must be `\0`-terminated.
    pub fn open(path: &[u8]) -> Result<Self> {
        let mut handle: bindings::ove_dir_t = core::ptr::null_mut();
        let rc = unsafe { bindings::ove_fs_opendir(&mut handle, path.as_ptr() as *const _) };
        Error::from_code(rc)?;
        Ok(Self { handle })
    }

    /// Read the next directory entry.
    ///
    /// Returns `Ok(None)` at end-of-directory. Different backends signal
    /// end-of-dir differently (error code vs. empty name), so both are
    /// treated as `Ok(None)`.
    pub fn read_entry(&mut self) -> Result<Option<DirEntry>> {
        let mut entry: bindings::ove_dirent = unsafe { core::mem::zeroed() };
        let rc = unsafe { bindings::ove_fs_readdir(self.handle, &mut entry) };
        if rc != 0 || entry.name[0] == 0 {
            return Ok(None);
        }
        Ok(Some(DirEntry { inner: entry }))
    }
}

impl Drop for Dir {
    fn drop(&mut self) {
        unsafe { bindings::ove_fs_closedir(self.handle) };
    }
}

// SAFETY: Dir wraps an opaque RTOS handle. Access is single-threaded by
// application design.
unsafe impl Send for Dir {}

fn cstr_to_slice(buf: &[u8]) -> &[u8] {
    let mut len = 0;
    while len < buf.len() && buf[len] != 0 {
        len += 1;
    }
    &buf[..len]
}
